### DISCUSSION TITLE
PySql里面如何实现跟mybatis一样效果的where子句

### DISCUSSION THREAD

**Discussion by liaohongxing** [rbatis/rbatis#301] [Q&A]
我想用PySql实现类似效果：
https://blog.csdn.net/PanTianTianO/article/details/82595107
http://www.qb5200.com/article/487597.html

```
第一，只有if标签有内容的情况下才会插入where子句；
第二，若子句的开通为 “AND” 或 “OR”，where标签会将它替换去除
```
这样就没有丑陋的 where 1=1 或者 where id != ''

```RUST
crud!(Article {});
impl_select_page!(Article{select_page(dto: &crate::domain::dto::ArticlePageDTO) =>
    "where:
        if dto.pid!=null:
          ` and pid= #{dto.pid}`
        if dto.cid!=null:
          ` and cid= #{dto.cid}`
    if !sql.contains('count'):
        ` order by created_at `"});

```

使用上面代码提示：

```
custom attribute panicked
message: [rbatis] parse py_sql fail!: E("[rbatis] unknow tag: `select count(1) as count from ${table_name} `where:")
```

请问我该如何做

**Comment by zhuxiujia** [comment]
> 我想用PySql实现类似效果： https://blog.csdn.net/PanTianTianO/article/details/82595107 http://www.qb5200.com/article/487597.html
> 
> ```
> 第一，只有if标签有内容的情况下才会插入where子句；
> 第二，若子句的开通为 “AND” 或 “OR”，where标签会将它替换去除
> ```
> 
> 这样就没有丑陋的 where 1=1 或者 where id != ''
> 
> ```rust
> crud!(Article {});
> impl_select_page!(Article{select_page(dto: &crate::domain::dto::ArticlePageDTO) =>
>     "where:
>         if dto.pid!=null:
>           ` and pid= #{dto.pid}`
>         if dto.cid!=null:
>           ` and cid= #{dto.cid}`
>     if !sql.contains('count'):
>         ` order by created_at `"});
> ```
> 
> 使用上面代码提示：
> 
> ```
> custom attribute panicked
> message: [rbatis] parse py_sql fail!: E("[rbatis] unknow tag: `select count(1) as count from ${table_name} `where:")
> ```
> 
> 请问我该如何做

试试

```
` where `
trim ' and ' : trim ' or ' :
        if dto.pid!=null:
          ` and pid= #{dto.pid}`
        if dto.cid!=null:
          ` and cid= #{dto.cid}`
```

**Comment by liaohongxing** [comment]
这个好像不满足if 条件都为空的情况下 ，自动去掉 where 的情况，会报错
